docs/enhancements/agent-late-binding.md

```diff
@@ -0,0 +1,125 @@
+
+- Enables the user story where the Discovery ISO is loaded at the factory, and then discovered at a later time.
+
+### Goals
+Note: The language here refers to CRDs, but also holds true for the corresponding REST resources
+
+- Make the InfraEnv and Agent resources independent from the ClusterDeployment. A single InfraEnv should have the potential to boot any number of hosts, whose Agents register with no relation to any particular ClusterDeployment.  An Agent may be assigned to a ClusterDeployment at a later time.
+
+- An Agent may be associated with a different ClusterDeployment at any time, unless it is part of an ongoing cluster installation.
+
+### Non-Goals
+Note: The language here refers to CRDs, but also holds true for the corresponding REST resources
+
+- Once a host boots from disk, booting the Discovery ISO again to return it to the collection of free Agents is considered to be a manual step.  There are several options that we will explore whereby this can be done automatically, but that discussion is out of scope for this enhancement.
+
+- At a later stage, Agents may be automatically assigned to clusters based on their discovered inventory and/or network configurations. This is future work and not in the scope of this enhancement.
+
+- For SaaS, there are discussions about changing the tenancy model to include organizations and not only users.  Those decisions are out of scope for this document.
+
+## Proposal
+
+There are two main (non-breaking) changes to the CRDs:
+1. ClusterRef in the InfraEnv Spec will no longer be mandatory.
+1. ClusterRef in the Agent Spec will no longer be mandatory.
+
+A Discovery ISO with no associated cluster means that hosts booted with that ISO will have agents that are not bound to any cluster.  Thus, Agent CRs can be created with no cluster binding.  It will still be possible to retain the previous behavior by providing a ClusterRef.  The multitenancy model will remain unchanged, with Agents being created in the namespace of the InfraEnv they booted from.
+
+The Agent's ClusterRef may now be deleted or changed dynamically, unless the Agent is part of an ongoing cluster installation, in which case the installation must first be cancelled (not currently possible via the k8s API).  If the host had started/completed its installation, a new Agent Condition would indicate that the user needs to boot the host into the Discovery ISO once
+again (as mentioned in the non-goals, this can be done automatically but is not in the scope of this enhancement).
+
+The REST API will require bigger changes:
+1. Move images to be a standalone resource, rather than part of the cluster resource (currently /clusters/{cluster_id}/downloads/image).  This removes the dependency of an image on a specific cluster, removes the current limitation of one image per cluster, and aligns with InfraEnv CRD.  The image resource will have an additional optional cluster-id property.
```

Comment:
Why have the option of associating an image to a cluster? Just to replicate the current behavior?

Reply:
Because (1) some users won't be interested in using late binding and it just introduces an extra step, and (2) to not break the current UI and test flows.